File: 740-delete-and-earn/delete-and-earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int ans = 0;
        sort(nums.begin(), nums.end());
        vector<int> acc(20000);
        for(int i=0; i<nums.size();){
            int sum = nums[i];
            int index = nums[i];
            i++;
            while(i < nums.size() && nums[i] == nums[i-1]){
                sum += nums[i];
                i++;
            }

            acc[index] = sum;
        }

        ans = maximumSum(acc);

        return ans;
    }

private:
    int maximumSum(vector<int> &acc){
        vector<int> dp(acc.size()+1, -1);
        return solve(acc, 0, dp);
    }
    int solve(vector<int> &acc, int index, vector<int> &dp){
        if (index >= acc.size()) return 0;

        if (dp[index] != -1) return dp[index];

        int take = acc[index] + solve(acc, index+2, dp);
        int not_take = solve(acc, index+1, dp);

        return dp[index] = max(take, not_take);
    }
};

File: 740-delete-and-earn/delete-and-earn.go
func deleteAndEarn(nums []int) int {
    max_ele := nums[0]
    for i := range nums {
        max_ele = max(max_ele, nums[i])
    }

    dp := make([]int, max_ele+5)
    for _, val := range nums {
        dp[val]+=val
    }

    for i := max_ele; i >= 0; i-- {
        dp[i] = max(dp[i] + dp[i+2], dp[i+1])
    }

    return dp[0]
}